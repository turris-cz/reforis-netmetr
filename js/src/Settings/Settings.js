/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import { ForisForm } from "foris";

import API_URLs from "API";
import AutostartForm from "./AutostartForm/AutostartForm";
import NetMetr from "../NetMetr";

Settings.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function Settings({ ws }) {
    return (
        <NetMetr>
            <ForisForm
                ws={ws}
                forisConfig={{
                    endpoint: API_URLs.settings,
                    wsModule: "netmetr",
                }}
                prepDataToSubmit={prepDataToSubmit}
                validator={validator}
            >
                <AutostartForm />
            </ForisForm>
        </NetMetr>
    );
}

function prepDataToSubmit(data) {
    delete data.sync_code;
    return data;
}

function validator(formData) {
    const error = {};
    if (formData.autostart_enabled && formData.hours_to_run.length === 0) {
        error.hours_to_run = "Hours must be selected";
    }
    return error.hours_to_run ? error : undefined;
}
