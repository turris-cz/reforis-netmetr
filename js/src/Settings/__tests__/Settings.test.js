/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, wait, fireEvent } from "foris/testUtils/customTestRender";
import { WebSockets } from "foris";
import diffSnapshot from "snapshot-diff";
import mockAxios from "jest-mock-axios";

import autostartFixture from "./__fixtures__/autostart";
import Settings from "../Settings";

describe("<Settings />", () => {
    let container;
    let asFragment;
    let getByText;
    let firstRender;

    beforeEach(async () => {
        const webSockets = new WebSockets();
        ({ container, asFragment, getByText } = render(
            <Settings ws={webSockets} />
        ));
        mockAxios.mockResponse({ data: autostartFixture });
        await wait(() => getByText("NetMetr"));
        firstRender = asFragment();
    });

    it("Should render component.", () => {
        expect(firstRender).toMatchSnapshot();
    });

    it("Should disable hours when disabled.", () => {
        fireEvent.click(getByText("Enabled"));
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("Should change form value when disabled.", () => {
        fireEvent.click(getByText("Enabled"));
        fireEvent.click(getByText("Save"));

        expect(mockAxios.post).toBeCalled();
        const data = {
            autostart_enabled: false,
            hours_to_run: [0, 1, 2, 3, 4],
        };
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/netmetr/api/settings",
            data,
            expect.anything()
        );
    });

    it("Should enable hours.", () => {
        fireEvent.click(getByText("7:00 AM"));
        fireEvent.click(getByText("12:00 PM"));
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("Should change form value when new hours are selected.", () => {
        fireEvent.click(getByText("7:00 AM"));
        fireEvent.click(getByText("12:00 PM"));
        fireEvent.click(getByText("Save"));

        expect(mockAxios.post).toBeCalled();
        const data = {
            autostart_enabled: true,
            hours_to_run: [0, 1, 2, 3, 4, 7, 12],
        };
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/netmetr/api/settings",
            data,
            expect.anything()
        );
    });
});
