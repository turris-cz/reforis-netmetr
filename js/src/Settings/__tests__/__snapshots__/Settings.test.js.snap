// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> Should disable hours when disabled. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -33,305 +33,10 @@
            >
              Enabled
            </label>
          </div>
        </div>
-       <h3>
-         Select autostart hours
-       </h3>
-       <div
-         class=\\"btn-group\\"
-       >
-         <button
-           class=\\"btn btn-sm btn-outline-info \\"
-           type=\\"button\\"
-         >
-           Select all
-         </button>
-         <button
-           class=\\"btn btn-sm btn-outline-info\\"
-           type=\\"button\\"
-         >
-           Unselect all
-         </button>
-       </div>
-       <div
-         class=\\"btn-group-toggle hour-buttons\\"
-         data-toggle=\\"buttons\\"
-       >
-         <label
-           class=\\"btn btn-outline-secondary active\\"
-           for=\\"2\\"
-         >
-           12:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"2\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary active\\"
-           for=\\"3\\"
-         >
-           1:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"3\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary active\\"
-           for=\\"4\\"
-         >
-           2:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"4\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary active\\"
-           for=\\"5\\"
-         >
-           3:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"5\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary active\\"
-           for=\\"6\\"
-         >
-           4:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"6\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"7\\"
-         >
-           5:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"7\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"8\\"
-         >
-           6:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"8\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"9\\"
-         >
-           7:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"9\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"10\\"
-         >
-           8:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"10\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"11\\"
-         >
-           9:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"11\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"12\\"
-         >
-           10:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"12\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"13\\"
-         >
-           11:00 AM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"13\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"14\\"
-         >
-           12:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"14\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"15\\"
-         >
-           1:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"15\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"16\\"
-         >
-           2:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"16\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"17\\"
-         >
-           3:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"17\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"18\\"
-         >
-           4:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"18\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"19\\"
-         >
-           5:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"19\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"20\\"
-         >
-           6:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"20\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"21\\"
-         >
-           7:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"21\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"22\\"
-         >
-           8:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"22\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"23\\"
-         >
-           9:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"23\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"24\\"
-         >
-           10:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"24\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-         <label
-           class=\\"btn btn-outline-secondary\\"
-           for=\\"25\\"
-         >
-           11:00 PM
-           <input
-             autocomplete=\\"off\\"
-             id=\\"25\\"
-             type=\\"checkbox\\"
-           />
-         </label>
-       </div>
-       <small>
-         <p
-           class=\\"text-muted\\"
-         >
-           The requested start time is approximate. The measurement will start about the selected hour (time is no exact due to load distribution).
-         </p>
-       </small>
        <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-md-3 col-lg-2\\""
`;

exports[`<Settings /> Should enable hours. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -134,11 +134,11 @@
              id=\\"8\\"
              type=\\"checkbox\\"
            />
          </label>
          <label
-           class=\\"btn btn-outline-secondary\\"
+           class=\\"btn btn-outline-secondary active\\"
            for=\\"9\\"
          >
            7:00 AM
            <input
              autocomplete=\\"off\\"
@@ -189,11 +189,11 @@
              id=\\"13\\"
              type=\\"checkbox\\"
            />
          </label>
          <label
-           class=\\"btn btn-outline-secondary\\"
+           class=\\"btn btn-outline-secondary active\\"
            for=\\"14\\"
          >
            12:00 PM
            <input
              autocomplete=\\"off\\""
`;

exports[`<Settings /> Should render component. 1`] = `
<DocumentFragment>
  <h1>
    NetMetr
  </h1>
  <p>
    NetMeter measures your internet parameters like upload, download and response time.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-3"
  >
    <form>
      <h2>
        Autostart
      </h2>
      <p>
        Enable Autostart to allow automatic measurement on specific daily hours that can be chosen.
      </p>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-switch custom-control-inline"
        >
          <input
            checked=""
            class="custom-control-input"
            id="1"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="1"
          >
            Enabled
          </label>
        </div>
      </div>
      <h3>
        Select autostart hours
      </h3>
      <div
        class="btn-group"
      >
        <button
          class="btn btn-sm btn-outline-info "
          type="button"
        >
          Select all
        </button>
        <button
          class="btn btn-sm btn-outline-info"
          type="button"
        >
          Unselect all
        </button>
      </div>
      <div
        class="btn-group-toggle hour-buttons"
        data-toggle="buttons"
      >
        <label
          class="btn btn-outline-secondary active"
          for="2"
        >
          12:00 AM
          <input
            autocomplete="off"
            id="2"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary active"
          for="3"
        >
          1:00 AM
          <input
            autocomplete="off"
            id="3"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary active"
          for="4"
        >
          2:00 AM
          <input
            autocomplete="off"
            id="4"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary active"
          for="5"
        >
          3:00 AM
          <input
            autocomplete="off"
            id="5"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary active"
          for="6"
        >
          4:00 AM
          <input
            autocomplete="off"
            id="6"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="7"
        >
          5:00 AM
          <input
            autocomplete="off"
            id="7"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="8"
        >
          6:00 AM
          <input
            autocomplete="off"
            id="8"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="9"
        >
          7:00 AM
          <input
            autocomplete="off"
            id="9"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="10"
        >
          8:00 AM
          <input
            autocomplete="off"
            id="10"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="11"
        >
          9:00 AM
          <input
            autocomplete="off"
            id="11"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="12"
        >
          10:00 AM
          <input
            autocomplete="off"
            id="12"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="13"
        >
          11:00 AM
          <input
            autocomplete="off"
            id="13"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="14"
        >
          12:00 PM
          <input
            autocomplete="off"
            id="14"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="15"
        >
          1:00 PM
          <input
            autocomplete="off"
            id="15"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="16"
        >
          2:00 PM
          <input
            autocomplete="off"
            id="16"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="17"
        >
          3:00 PM
          <input
            autocomplete="off"
            id="17"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="18"
        >
          4:00 PM
          <input
            autocomplete="off"
            id="18"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="19"
        >
          5:00 PM
          <input
            autocomplete="off"
            id="19"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="20"
        >
          6:00 PM
          <input
            autocomplete="off"
            id="20"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="21"
        >
          7:00 PM
          <input
            autocomplete="off"
            id="21"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="22"
        >
          8:00 PM
          <input
            autocomplete="off"
            id="22"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="23"
        >
          9:00 PM
          <input
            autocomplete="off"
            id="23"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="24"
        >
          10:00 PM
          <input
            autocomplete="off"
            id="24"
            type="checkbox"
          />
        </label>
        <label
          class="btn btn-outline-secondary"
          for="25"
        >
          11:00 PM
          <input
            autocomplete="off"
            id="25"
            type="checkbox"
          />
        </label>
      </div>
      <small>
        <p
          class="text-muted"
        >
          The requested start time is approximate. The measurement will start about the selected hour (time is no exact due to load distribution).
        </p>
      </small>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
